torch/sparse/__init__.py

```diff
@@ -491,3 +492,119 @@ def test_mth(*args, **kwargs):
+elements (in addition to carrying a element value), define a masked-in
+predicate that enables computations with such elements. The
+unspecified sparse tensor elements define a masked-out predicate that
+disables the corresponding computations. This assumption defines the
+so-called masked semantics.
+
+The following options exists to manage sparse semantics for autograd:
+
+1. Using a context manager:
+
+   .. code:: python
+
+       run_my_model()  # assuming the default sparse semantics
+
+       with torch.sparse.sparse_semantics(False):
+           run_my_model()  # assuming masked semantics
+
+2. Using a procedural approach:
+
+   .. code:: python
+
+       prev_sparse_semantics = torch.sparse.sparse_semantics.is_enabled()
+       torch.sparse.sparse_semantics.enable()
+
+       run_my_model()  # assuming sparse semantics
+
+       if not prev_sparse_semantics:
+           torch.sparse.sparse_semantics.disable()
+
+3. Using function decoration:
+
+   .. code:: python
+
+       @torch.sparse.sparse_semantics(False)
+       def run_my_model():
+           ...
+
+       run_my_model()  # assuming masked semantics
+
+    """
+    @staticmethod
+    def is_enabled():
+        r"""Returns True if the sparse semantics is enabled.
+
+.. note::
+
+    Use :func:`torch.sparse.sparse_semantics.enable` or
+    :func:`torch.sparse.sparse_semantics.disable` to
+    manage the state of the sparse semantics.
+        """
+        return torch._C._sparse_semantics()
+
+    @staticmethod
+    def enable():
+        r"""Enable sparse semantics for autograd.
+
+.. note::
+
+    By default, the sparse semantics is enabled. Use
+    :func:`torch.sparse.sparse_semantics.is_enabled` to retrieve the
+    current state of sparse semantics.
+
+.. note::
+
+    The sparse semantics flag is effective in both Python and ATen.
+
+    The flag can be locally overridden by the
+    :func:`torch.sparse.sparse_semantics` context manager.
+        """
+        torch._C._set_sparse_semantics(True)
+
+    @staticmethod
+    def disable():
+        r"""Disable sparse tensor invariants checking in sparse tensor constructors.
+
+See :func:`torch.sparse.sparse_semantics.enable` for more information.
+        """
+        torch._C._set_sparse_semantics(False)
+
+    # context manager support
+    def __init__(self, enable=True):
+        self.state = enable
+        self.saved_state : Optional[bool] = None
+
+    def __enter__(self):
+        if self.saved_state is not None:
+            raise RuntimeError('This context manager instance is already activated.'
+                               ' Use a different context manager instance for context nesting.')
+        self.saved_state = self.is_enabled()
+        torch._C._set_sparse_semantics(self.state)
+
+    def __exit__(self, type, value, traceback):
+        assert self.saved_state is not None
+        torch._C._set_sparse_semantics(self.saved_state)
+        self.saved_state = None
+
+    # decorator support
```

Comment:
@pearu Wouldn't it just be cleaner to have it inherit from contextlib.ContextDecorator?